### PR TITLE
add coat

### LINKED ISSUE #426: [Models][330][transformer]coat
If this is your first time, please read our contributor guidelines: https://gitee.com/mindspore/mindspore/blob/master/CONTRIBUTING.md

**Is your feature request related to a problem? Please describe.**
https://arxiv.org/abs/2104.06399
https://github.com/mlpc-ucsd/CoaT

**Describe the solution you'd like**
model.py
readme
config

### PR DESCRIPTION
**Pull request by SunJuil-Ty**
Thank you for your contribution to the MindCV repo.
Before submitting this PR, please make sure:

- [ ] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved terminology
- [ ] You have added unit tests

## Motivation

(Write your motivation for proposed changes here.)

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)
closes #426

### CODE REVIEW

File: configs/coat/coat_lite_mini.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# system config
+mode: 0
+enable_modelarts: True
```
Comment (SamitHuang) [inline configs/coat/coat_lite_mini.yaml:3]: remove

File: configs/coat/README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+## Results
+
+Our reproduced model performance on ImageNet-1K is reported as follows.
+
+<div align="center">
+
+| Model           | Context   |  Top-1 (%)  | Params (M) | Recipe                                                                                   | Weight                                                                            |
+|-----------------|-----------|-------|------------|------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------|
+| coat_lite_tiny | D910x8-G | 77.35 | 5.7       | [yaml](https://github.com/mindspore-lab/mindcv/blob/main/configs/coat/coat_lite_tiny.yaml) | [weights](https://download.mindspore.cn/toolkits/mindcv/coat/coat_lite_tiny_acc77.35_bs64_8p.ckpt) |
```
Comment (SamitHuang) [inline configs/coat/README.md:17]: 权重命名有问题，请参考其他模型。

File: configs/coat/README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+**Note:** As the global batch size (batch_size x num_devices) is an important hyper-parameter, it is recommended to keep the global batch size unchanged for reproduction or adjust the learning rate linearly to a new global batch size.
+
+- Standalone Training
+
+If you want to train or finetune the model on a smaller dataset without distributed training, please run:
+
+```
+# standalone training on a CPU/GPU/Ascend device
+python train.py --config configs/rexnet/rexnet_x09.yaml --data_dir /path/to/dataset --distribute False
```
Comment (SamitHuang) [inline configs/coat/README.md:62]: config名错误

File: configs/coat/coat_lite_mini.yaml (hunk 1 above)
Comment (SunJuil-Ty) [reply configs/coat/coat_lite_mini.yaml]: 已修正

File: configs/coat/README.md (hunk 2 above)
Comment (SunJuil-Ty) [reply configs/coat/README.md]: 已改正

File: configs/coat/README.md (hunk 3 above)
Comment (SunJuil-Ty) [reply configs/coat/README.md]: 已改正

File: configs/coat/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+Co-Scale Conv-Attentional Image Transformer (CoaT) is a Transformer-based image classifier equipped with co-scale and conv-attentional mechanisms. First, the co-scale mechanism maintains the integrity of Transformers' encoder branches at individual scales, while allowing representations learned at different scales to effectively communicate with each other. Second, the conv-attentional mechanism is designed by realizing a relative position embedding formulation in the factorized attention module with an efficient convolution-like implementation. CoaT empowers image Transformers with enriched multi-scale and contextual modeling capabilities.
+
+## Results
+
+Our reproduced model performance on ImageNet-1K is reported as follows.
+
+<div align="center">
+
+| Model           | Context   |  Top-1 (%)  | Params (M) | Recipe                                                                                   | Weight                                                                            |
```
Comment (Songyuanwei) [inline configs/coat/README.md:15]: 缺少Top-5 (%) ，Params 保存小数点后两位数字

File: configs/coat/coat_lite_mini.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# system config
```
Comment (Songyuanwei) [inline configs/coat/coat_lite_mini.yaml:1]: system config -> system 去掉config， 下同

File: configs/coat/coat_lite_mini.yaml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+mode: 0
+distribute: True
+num_parallel_workers: 8
+val_while_train: True
+
+# dataset config
+dataset: 'imagenet'
+data_dir: './imagenet/'
```
Comment (Songyuanwei) [inline configs/coat/coat_lite_mini.yaml:9]: './imagenet/' -> '/path/to/imagenet'

File: configs/coat/coat_lite_mini.yaml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+shuffle: True
+dataset_download: False
+batch_size: 64
+drop_remainder: True
+
+# Augmentation config
```
Comment (Songyuanwei) [inline configs/coat/coat_lite_mini.yaml:15]: Augmentation config -> augmentation, 首字母小写，去掉config， 下同

File: configs/coat/coat_lite_mini.yaml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ckpt_save_dir: './ckpt/'
+dataset_sink_mode: True
+amp_level: 'O2'
+
+# loss config
+loss: 'CE'
+label_smoothing: 0.1
+
+# lr scheduler config
+scheduler: 'warmup_cosine_decay'
```
Comment (Songyuanwei) [inline configs/coat/coat_lite_mini.yaml:45]: 参数更正： 'warmup_cosine_decay' -> 'cosine_decay'

File: configs/coat/coat_lite_mini.yaml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+num_cycles: 4
+cycle_decay: 1.0
+
+# optimizer config
+opt: 'adamw'
+weight_decay: 0.025
+filter_bias_and_bn: True
+loss_scale: 1024
+use_nesterov: False
+dynamic_loss_scale: True
```
Comment (Songyuanwei) [inline configs/coat/coat_lite_mini.yaml:60]: 文件末尾缺少一空行

File: mindcv/models/coat.py [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+
+
+__all__ = [
+    "coat_tiny",
+    "coat_mini",
+    "coat_small",
+    "coat_lite_tiny",
+    "coat_lite_mini",
+    "coat_lite_small",
+    "coat_lite_medium"
```
Comment (Songyuanwei) [inline mindcv/models/coat.py:29]: 末尾缺少逗号"coat_lite_medium"-> "coat_lite_medium",

File: mindcv/models/coat.py [hunk 11]
```diff
@@ -0,0 +1,791 @@
...
+        'first_conv': 'patch_embed.proj', 'classifier': 'head',
+        **kwargs
+    }
+
+
+default_cfgs = {
+    'coat_tiny': _cfg(url=''),
+    'coat_mini': _cfg(url=''),
+    'coat_small': _cfg(url=''),
+    'coat_lite_tiny': _cfg(url=''),
```
Comment (Songyuanwei) [inline mindcv/models/coat.py:51]: 缺少coat_lite_tiny 和coat_lite_mini的url链接

File: mindcv/models/coat.py [hunk 12]
```diff
@@ -0,0 +1,791 @@
...
+    default_cfg = default_cfgs['coat_lite_medium']
+    model = CoaT(in_channels=in_channels, num_classes=num_classes,
+                 patch_size=4, embed_dims=[128, 256, 320, 512],
+                 serial_depths=[3, 6, 10, 8], parallel_depth=0,
+                 num_heads=8, mlp_ratios=[4, 4, 4, 4], **kwargs)
+    model.default_cfg = default_cfg
+
+    if pretrained:
+        load_pretrained(model, default_cfg, num_classes=num_classes, in_channels=in_channels)
+    return model
```
Comment (Songyuanwei) [inline mindcv/models/coat.py:791]: 文件末尾缺少一空行

File: configs/coat/coat_lite_mini.yaml (hunk 6 above)
Comment (SunJuil-Ty) [reply configs/coat/coat_lite_mini.yaml]: 已修改

File: configs/coat/coat_lite_mini.yaml (hunk 9 above)
Comment (SunJuil-Ty) [reply configs/coat/coat_lite_mini.yaml]: 已添加

File: mindcv/models/coat.py (hunk 10 above)
Comment (SunJuil-Ty) [reply mindcv/models/coat.py]: 已添加逗号

File: mindcv/models/coat.py (hunk 11 above)
Comment (SunJuil-Ty) [reply mindcv/models/coat.py]: 已添加url

File: mindcv/models/coat.py (hunk 12 above)
Comment (SunJuil-Ty) [reply mindcv/models/coat.py]: 已添加

File: configs/coat/README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+## Results
+
+Our reproduced model performance on ImageNet-1K is reported as follows.
+
+<div align="center">
+
+| Model           | Context   |  Top-1 (%)  |  Top-5 (%)  | Params (M) | Recipe                                                                                   | Weight                                                                            |
+|-----------------|-----------|-------|------------|------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------|----------------------------------------------------------------------------------|
+| coat_lite_tiny | D910x8-G | 77.35 | 93.43 | 5.7       | [yaml](https://github.com/mindspore-lab/mindcv/blob/main/configs/coat/coat_lite_tiny.yaml) | [weights](https://download.mindspore.cn/toolkits/mindcv/coat/coat_lite_tiny_fa7bf894.ckpt) |
```
Comment (Songyuanwei) [inline configs/coat/README.md:17]: params 参数保留小数点后两位。

File: configs/coat/coat_lite_mini.yaml [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+num_cycles: 4
+cycle_decay: 1.0
+
+# optimizer
+opt: 'adamw'
+weight_decay: 0.025
+filter_bias_and_bn: True
+loss_scale: 1024
+use_nesterov: False
+dynamic_loss_scale: True
```
Comment (Songyuanwei) [inline configs/coat/coat_lite_mini.yaml:60]: 参数更正 dynamic_loss_scale: True  -> loss_scale_type: "dynamic"
增加         drop_overflow_update: True
这两个必须一起出现

**Review by geniuspatrick — APPROVED** [review]
